reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ThrottleTest.kt

```diff
@@ -0,0 +1,110 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.testutils.TestObservable
+import com.badoo.reaktive.testutils.hasOnNext
+import com.badoo.reaktive.testutils.test
+import com.badoo.reaktive.testutils.values
+import com.badoo.reaktive.utils.atomicreference.AtomicReference
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+
+class ThrottleTest : UpstreamDownstreamGenericTests by UpstreamDownstreamGenericTests<Unit>({ throttle(0L) }) {
+
+    private val timeMillis = AtomicReference(0L)
+    private val upstream = TestObservable<Int>()
+    private val observer = upstream.throttle(100L, timeMillis::value).test()
+
+    @Test
+    fun emits_first_value_WHEN_current_time_is_0L() {
+        emit(0)
+
+        assertValues(0)
+    }
+
+    @Test
+    fun emits_first_value_WHEN_current_time_is_less_than_window() {
+        setTime(99L)
+        emit(0)
+
+        assertValues(0)
+    }
+
+    @Test
+    fun emits_first_value_WHEN_current_time_is_equals_to_window() {
+        setTime(100L)
+        emit(0)
+
+        assertValues(0)
+    }
+
+    @Test
+    fun emits_first_value_WHEN_current_time_is_more_than_window() {
+        setTime(101L)
+        emit(0)
+
+        assertValues(0)
+    }
+
+    @Test
+    fun does_not_emit_WHEN_timeout_is_not_passed() {
+        emit(0)
+        observer.reset()
+        emit(1)
+        setTime(20L)
+        emit(2)
+        emit(3)
+        setTime(60L)
+        emit(4)
+        setTime(99L)
+        emit(5)
+        emit(6)
+
+        assertFalse(observer.hasOnNext)
+    }
+
+    @Test
+    fun emits_WHEN_timeout_is_passed() {
+        emit(0)
+        setTime(99L)
+        emit(1)
+        setTime(100L)
+        observer.reset()
+
+        emit(2)
+
+        assertValues(2)
+    }
+
+    @Test
+    fun emits_correct_values_WHEN_complex_series() {
```

Comment:
Not sure what this is testing.

Reply:
It sends some tricky series of values through upstream and verifies that output is correct

Reply:
I think it would be good to be explicit about how it is complex. 
In the future we may want tests to cover specific complex scenarios, it may mix us up what is and isn't already covered.

Reply:
@mk-jones I can't find any words to describe this particular case :-)